pkg/utils/grpc.go

```diff
@@ -27,44 +46,136 @@ func NewGrpc(serverType string, port int, certPem, keyPem, rootCertPem []byte) (
 
 	log.Debug().Msgf("Parameters for %s gRPC server: MaxConcurrentStreams=%d;  KeepAlive=%+v", serverType, maxStreams, streamKeepAliveDuration)
 
+	s := &Server{
+		name:     serverType,
+		cm:       cm,
+		certName: certName,
+	}
+
 	grpcOptions := []grpc.ServerOption{
 		grpc.MaxConcurrentStreams(maxStreams),
 		grpc.KeepaliveParams(keepalive.ServerParameters{
 			Time: streamKeepAliveDuration,
 		}),
 	}
 
-	mutualTLS, err := setupMutualTLS(false, serverType, certPem, keyPem, rootCertPem)
-	if err != nil {
-		log.Error().Err(err).Msg("Error setting up mutual tls for GRPC server")
-		return nil, nil, err
+	// #nosec G402: TLS MinVersion too low
+	tlsConfig := tls.Config{
+		InsecureSkipVerify: false,
+		ServerName:         serverType,
+		ClientAuth:         tls.RequireAndVerifyClientCert,
+		MinVersion:         constants.MinTLSVersion,
+		GetConfigForClient: func(*tls.ClientHelloInfo) (*tls.Config, error) {
+			// use lock to prevent concurrent updates and reads to the tls config
+			s.mu.Lock()
+			defer s.mu.Unlock()
+			return &s.config, nil
+		},
 	}
+	mutualTLS := grpc.Creds(credentials.NewTLS(&tlsConfig))
 	grpcOptions = append(grpcOptions, mutualTLS)
 
-	return grpc.NewServer(grpcOptions...), lis, nil
+	s.server = grpc.NewServer(grpcOptions...)
+	return s, lis, nil
+}
+
+// GetServer returns the gRPC server
+func (s *Server) GetServer() *grpc.Server {
+	return s.server
+}
+
+func (s *Server) initConfig() error {
+	grpcCert, err := s.cm.IssueCertificate(
+		s.certName,
+		certificate.Internal,
+		certificate.FullCNProvided())
+	if err != nil {
+		return err
+	}
+
+	certif, err := tls.X509KeyPair(grpcCert.GetCertificateChain(), grpcCert.GetPrivateKey())
+	if err != nil {
+		return fmt.Errorf("failed loading Certificate (%+v) and Key (%+v) PEM files", grpcCert.GetCertificateChain(), grpcCert.GetPrivateKey())
+	}
+
+	certPool := x509.NewCertPool()
+
+	// Load the set of Root CAs
+	if ok := certPool.AppendCertsFromPEM(grpcCert.GetTrustedCAs()); !ok {
+		return fmt.Errorf("failed to append client certs")
+	}
+
+	// use lock to prevent concurrent updates and reads to the tls config
+	s.mu.Lock()
+	// #nosec G402: TLS MinVersion too low
+	s.config = tls.Config{
+		InsecureSkipVerify: false,
+		ServerName:         s.name,
+		ClientAuth:         tls.RequireAndVerifyClientCert,
+		Certificates:       []tls.Certificate{certif},
+		ClientCAs:          certPool,
+		MinVersion:         constants.MinTLSVersion,
+	}
+	s.mu.Unlock()
+
+	return nil
+}
+
+func (s *Server) watchCertRotations(ctx context.Context) error {
+	// listen for certificate rotation first, so we don't miss any events
+	certRotationChan, unsubscribeRotation := s.cm.SubscribeRotations(s.certName)
+	defer unsubscribeRotation()
```

Comment:
won't this unsubscribe when this func returns?

Reply:
Yes, that was accidentally included here. I highlighted this issue in [this conversation], but included it anyway :)(https://github.com/openservicemesh/osm/pull/4996#discussion_r948311899).